### PR TITLE
insights: Add endpoint for getting related insights inline

### PR DESCRIPTION
**Pull request by CristinaBirkel**
Closes https://github.com/sourcegraph/sourcegraph/issues/39121

## Description

This is a prototype to enable inline links to insights. We'll be able to see if devs find this feature useful as part of their core workflow. The main caveat here is that this is slow. In my testing it takes roughly 15s to match against 50 series. It's slow because it's making that many search queries as part of aggregating the results. We'll need to keep an eye on performance degradation on the instance when this feature is eventually turned on. We only expect this endpoint to be used on the file view--not on the main search page--and to be hidden behind a feature flag.

## Test plan
You can test this out locally with a query similar to:
```
query {
    relatedInsightsInline(input: {
        file: "enterprise/internal/insights/query/querybuilder/builder.go"
        revision: "a84fde5e35eed41060acb6aa46b842af32133027",
        repo: "github.com/sourcegraph/sourcegraph"}) 
    {
        viewId, title, lineNumbers
    }
}
```

### CODE REVIEW

**Comment by leonore** [comment]
just FYI there's a couple changes I've made in #39249 that you might want to pull into this PR instead for cleanliness? (repo input is never nil, renamed some graphql stuff)

File: enterprise/internal/insights/resolvers/related_insights_resolvers.go [hunk 1]
```diff
@@ -41,24 +41,21 @@ func (r *Resolver) RelatedInsightsInline(ctx context.Context, args graphqlbacken
 		}
 		mr := *metadataResult
 		if len(mr.Errors) > 0 {
-			log15.Warn("related insights errors", mr.Errors)
+			r.logger.Warn("related insights errors", log.Strings("errors", mr.Errors))
```
Comment (leonore) [inline enterprise/internal/insights/resolvers/related_insights_resolvers.go:44]: thanks, I was trying to use the logger but couldn't figure out what second field it was expecting and forgot to look at the docs and change it!

**Review by leonore — APPROVED** [review]

**Comment by coury-clark** [comment]
I think that we need to be somewhat concerned about the possibility for abuse here given these internal requests would bypass any kind of gateway level rate limiting or controls an instance may have. I think we should only initialize this resolver if some site level flag is enabled.

**Comment by CristinaBirkel** [comment]
@coury-clark The feature will be behind a feature-flag (default to off,) on the frontend, and I think we'll be able to communicate the potential performance impact with any customers wanting to turn this on. I do agree that there is potential for abuse at the API level, but I wonder if we need to go so far as to disable the endpoint? If we do disable it entirely, I would say it should be behind the same feature flag the the frontend is going to use.

I'll add a separate ticket for this though, to do this week.

**Comment by coury-clark** [comment]
@CristinaBirkel If this query is going to be potentially abusable (which is clear it is), we need to ship it behind an admin control instead of a user control. We can't risk this landing on a customer instance. Frontend flags will not be sufficient since they can be set by any arbitrary user.

**Comment by CristinaBirkel** [comment]
@coury-clark In that case, I guess it would be a two step process for a customer to demo this. First they'd need to get an admin to turn the API on. And then they would have the ability to turn the feature flag on and off as needed. This would prevent any customers who don't want to try this at all from ever having any issues with it.

@Joelkw any thoughts here? I just want to quickly verify that this doesn't add too much friction into the processing of getting customers to demo the prototype. If we don't lock the API behind a site setting, it's possible that a user could call the API directly and negatively impact the performance of the whole sourcegraph instance. Practically speaking, this would either have to be malicious, or else the user is exploring the API and writes some code that calls these new endpoints over and over again for some reason. It could happen!

**Comment by Joelkw** [comment]
@CristinaBirkel thanks for checking! I understand. Let's ship it in the 3.42 release behind the Admin flag as well, then, and we'll see what we learn from admins who do turn it on. Thanks for checking.

**Pull request merged**